lib/widgets/action_sheet.dart

```diff
@@ -12,51 +15,210 @@ import 'actions.dart';
+                        bottom: null,
+                        child: Container(
+                          height: math.min(scrollController.position.extentBefore, 16),
+                          decoration: BoxDecoration(
+                            gradient: LinearGradient(
+                              begin: Alignment.topCenter,
+                              end: Alignment.bottomCenter,
+                              colors: [
+                                designVariables.bgContextMenu,
+                                designVariables.bgContextMenu.withOpacity(0),
+                              ],
+                            ),
+                          ),
+                        ),
+                      );
+                    }
+                  ),
+                  // Serves as the bottom shadow:
+                  //   https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=3483-42291&t=3rYe12EFTQL3wfZB-1
+                  _ScrollControllerBuilder(
+                    scrollController: scrollController,
+                    builder: (_, scrollController) {
+                      final designVariables = DesignVariables.of(context);
+                      return Positioned.fill(
+                        top: null,
+                        bottom: math.max(8 - scrollController.position.extentAfter, 0),
+                        child: Container(
+                          height: math.min(scrollController.position.extentAfter, 8),
+                          decoration: BoxDecoration(
+                            gradient: LinearGradient(
+                              begin: Alignment.topCenter,
+                              end: Alignment.bottomCenter,
+                              colors: [
+                                designVariables.bgContextMenu.withOpacity(0),
+                                designVariables.bgContextMenu,
+                              ],
+                            ),
+                          ),
+                        ),
+                      );
+                    }
+                  ),
+                ],
+              ),
+            ),
+            const MessageActionSheetCancelButton(),
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+class _ScrollControllerBuilder extends StatelessWidget {
+  const _ScrollControllerBuilder({required this.scrollController, required this.builder});
+
+  final ScrollController scrollController;
+  final Widget Function(BuildContext, ScrollController) builder;
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: Future.microtask(() => true),
+      builder: (context, snapshot) {
+        if (!snapshot.hasData) {
+          return const SizedBox.shrink();
+        }
+        return ListenableBuilder(
+          listenable: scrollController,
+          builder: (context, __) => builder(context, scrollController),
+        );
+      });
```

Comment:
The top and bottom shadows appear only when there is content above or under the available height of the `ScrollView` so we listen to `ScrollController` and act upon its `position.extentBefore` and `position.extentAfter`. `FutureBuilder` is there to wait for `ScrollController` to be attached to the `SingleChildScrollView` so we can use those properties. `ListenableBuilder` is necessary for efficiently listening to `ScrollController` and redrawing a portion of the UI instead of all of the UI in the `build` method. So the junk will only happen to those small portions of the UI.

> In a search just now, I found something called [`ShaderMask`](https://api.flutter.dev/flutter/widgets/ShaderMask-class.html); could that be helpful?

I don't think it's going to be helpful in this specific case as we have to wrap the whole `SingleChildScrollView` with `ShaderMask` and now that we want to show those shadows based on the conditions mentioned above, the whole `SingleChildScrollView` has to be redrawn each time, which is inefficient.